### PR TITLE
use static graph to reduce restore times

### PR DESCRIPTION
**Pull request by jmarolf**

### CODE REVIEW

File: Restore.cmd [hunk 1]
```diff
@@ -1,2 +1,2 @@
 @echo off
-powershell -ExecutionPolicy ByPass -NoProfile -command "& """%~dp0eng\build.ps1""" -restore %*"
+powershell -ExecutionPolicy ByPass -NoProfile -command "& """%~dp0eng\build.ps1""" -restore  /p:RestoreUseStaticGraphEvaluation=true %*"
```
Comment (tmat) [inline Restore.cmd:2]: This should be added here: https://github.com/dotnet/roslyn/blob/master/eng/build.ps1#L264
`*.cmd` files are meant to be just trivial stubs dispatching to build.ps1. No logic/settings should be specified here.

**Comment by jaredpar** [comment]
Can we link to documentation that describes this feature? Particularly what restrictions come along with having this?

**Comment by jmarolf** [comment]
@jaredpar this was the motivating issue for the nuget team: https://github.com/NuGet/Home/issues/8791. You can see that issue for more background

The drawbacks are that if your nuget restore requires a design-time build to be correct (your install a nuget package which includes props or targets at build time which then add more package references that need to be restored).

This is not the case today for roslyn and I would argue that our build should not be in a state where we don't know all of our package references unless we do multiple design-time builds.

**Comment by jaredpar** [comment]
@jmarolf 

> The drawbacks are that if your nuget restore requires a design-time build to be correct (...)

You didn't specify the drawbacks, only that there are drawbacks :smile:

> This is not the case today for roslyn and I would argue that our build should not be in a state where we don't know all of our package references unless we do multiple design-time builds.

Agreed. But what happens if we violate this: does restore fail, silently drop back to non-static graph, etc ...

**Comment by jmarolf** [comment]
> Agreed. But what happens if we violate this: does restore fail, silently drop back to non-static graph, etc ...

My understanding is that restore fails in this case

File: Restore.cmd (hunk 1 above)
Comment (jmarolf) [reply Restore.cmd]: thanks, I'll move it to Build.ps1

**Comment by jaredpar** [comment]
Have you measured that this is actually faster?

**Comment by jmarolf** [comment]
> Have you measured that this is actually faster?

Restoring master today:
image


With this change:
image

**Comment by jaredpar** [comment]
Hmm, seeing ~8 seconds in CI. That' why I was asking. Guessing it's just slower machines. 

Oh well. Not a regression.

**Comment by jmarolf** [comment]
I would not expect as large of a change for 'cold' restores since the same amount of work is happening. For incremental restores like the one I show up above (where this is actually no work to do) things should be significantly faster since far less work needs to be done to determine that a restore is unnecessary.

**Comment by jaredpar** [comment]
please add a comment as to when we can switch to the fast restore using `msbuild`. That way we get the benefit locally as most restore goes through `msbuild`

**Comment by RikkiGibson** [comment]
@jmarolf can we turn this on in master now?

**Comment by jmarolf** [comment]
> can we turn this on for all builds in master now?

@RikkiGibson once this PR goes in https://github.com/dotnet/roslyn/pull/47165  we can turn it on for all restores

**Comment by jaredpar** [comment]
How does that PR help?

**Pull request merged**